 |  **Type** |  **Type** |  **Element** |  **Panel** |  **Concept** |  **Status** |  **Code** |
 |  :---:  |  :---:  |  :---:  |  :---:  |  :---:  |  :---:  |  :---:  |  
 |  Figure |  Main text |  Figure 1 |  a |  map of samples by environment plus workflow for AMPSphere |  (Map done) |  plot-metadata.py |  
 |  Figure |  Main text |  Figure 1 |  b |  quality distribution of homologs by mapped database |  Done |  quality_homologs_by_db.py |  
 |  Figure |  Main text |  Figure 1 |  c |  heatmap of amp overlap in higher environments |  Done |  henvo_alt.py |  
 |  Figure |  Main text |  Figure 1 |  d |  heatmap of amp overlap in mammal guts (min 100 samples per env.) |  Done |  hmammal_guts.py |  
 |  Figure |  Main text |  Figure 1 |  e |  heatmap of amp overlap in human body sites |  Done |  hbody.py |  
 |  Figure |  Main text |  Figure 2 |  a |  taxonomy affiliation of AMPSphere candidates (distribution of taxonomy assignment results) |  Done |  taxonomy.py |  
 |  Figure |  Main text |  Figure 2 |  b |  Tree of Prevotella species and the AMP contents per species and environment they were isolated from |  Done |  prevotella_amp.py |  
 |  Figure |  Main text |  Figure 2 |  c |  Frequency distribution of families by the number of genes their candidates are affiliated to |  Done |  family_sizes.py |  
 |  Figure |  Main text |  Figure 2 |  d |  Bar chart of number of genes coding for amps - x axis number of genes coding for a single AMP and y axis the number of candidates |  Done |  ugenes.py |  
 |  Figure |  Main text |  Figure 2 |  e |  bar chart with the proportion of accessory, shell and core genes and families, quality-controlled and in the entire set |  Done |  core.py |  
 |  Figure |  Main text |  Figure 3 |  a |  bimodal distribution of hits start site in the target sequences, kdeplot from the position start in the larger protein as percent |  Done |  plot_gmgc.py |  
 |  Figure |  Main text |  Figure 3 |  b |  mutations behind the origin of the peptide HG4 |  (Done but needs improvement) |  - |  
 |  Figure |  Main text |  Figure 3 |  c |  volcano plot with the eggNOG-OG enrichment in the homologs to GMGC, significance shown by color |  Not done |  ? |  
 |  Figure |  Main text |  Figure 3 |  d |  selection pressure |  Not done |  ? |  
 |  Figure |  Main text |  Figure 4 |  ? |  genome context |  Not done |  ? |  
 |  Figure |  Main text |  Figure 5 |  a |  box plot showing the distribution of AMPs per assembled Mbp in host- vs. non-host-associated samples |  Done |  plot-metadata.py |  
 |  Figure |  Main text |  Figure 5 |  b |  FMT workflow |  Not done |  - |  
 |  Figure |  Main text |  Figure 5 |  c |  conceptual results from FMT story showing the AMP effects in the outcome |  Not done |  ? |  
 |  Figure |  Main text |  Figure 6 |  ? |  Activity |  Not done |  ? |  
 |  Figure |  Supp. |  Figure S1 |  a |  kde plot showing comparison of length of peptides in AMPSphere, DRAMP and Macrel training set |  Done |  jelena_s.py |  
 |  Figure |  Supp. |  Figure S1 |  b |  kde plot showing comparison of percent composition of small residues in peptides from AMPSphere, DRAMP and Macrel training set |  Done |  jelena_s.py |  
 |  Figure |  Supp. |  Figure S1 |  c |  kde plot showing comparison of percent composition of basic residues in peptides from AMPSphere, DRAMP and Macrel training set |  Done |  jelena_s.py |  
 |  Figure |  Supp. |  Figure S1 |  d |  kde plot showing comparison of isoelectric point in peptides from AMPSphere, DRAMP and Macrel training set |  Done |  jelena_s.py |  
 |  Figure |  Supp. |  Figure S1 |  e |  kde plot showing comparison of charge at pH 7.0 in peptides from AMPSphere, DRAMP and Macrel training set |  Done |  jelena_s.py |  
 |  Figure |  Supp. |  Figure S1 |  f |  kde plot showing comparison of aliphatic index in peptides from AMPSphere, DRAMP and Macrel training set |  Done |  jelena_s.py |  
 |  Figure |  Supp. |  Figure S1 |  g |  kde plot showing comparison of instability index in peptides from AMPSphere, DRAMP and Macrel training set |  Done |  jelena_s.py |  
 |  Figure |  Supp. |  Figure S1 |  h |  kde plot showing comparison of Boman index in peptides from AMPSphere, DRAMP and Macrel training set |  Done |  jelena_s.py |  
 |  Figure |  Supp. |  Figure S1 |  i |  kde plot showing comparison of H-moment in peptides from AMPSphere, DRAMP and Macrel training set |  Done |  jelena_s.py |  
 |  Figure |  Supp. |  Figure S2 |  a |  scatter plot showing assembled Mbp versus AMPs |  Done |  suptable_s1.py |  
 |  Figure |  Supp. |  Figure S2 |  b |  scatter plot showing assembly N50 as Mbp versus AMPs |  Done |  suptable_s1.py |  
 |  Figure |  Supp. |  Figure S2 |  c |  scatter plot showing predict smORFs in millions versus AMPs |  Done |  suptable_s1.py |  
 |  Figure |  Supp. |  Figure S2 |  d |  scatter plot showing predicted ORFs in millions versus predicted smORFs in millions |  Done |  suptable_s1.py |  
 |  Figure |  Supp. |  Figure S2 |  e |  scatter plot showing the amps per Gbp assembled per genus â€“ correlation close to 1 |  Done |  norm_w_bps_per_taxon.py |  
 |  Figure |  Supp. |  Figure S2 |  f |  boxplot showing the difference of assembled base pairs in genera with presence and absence of amps |  Done |  norm_w_bps_per_taxon.py |  
 |  Figure |  Supp. |  Figure S3 |  - |  3 panels (1 for each clustering level) in horizontal showing identity of matches vs. log(E-value) for the cluster validation |  Done |  displotter.py |  
 |  Figure |  Supp. |  Figure S4 |  - |  Box plot showing the normalized number of AMPs per assembled Mbp per environment colored by host- and non-host-associated |  Done |  plot-metadata.py |  
 |  Table |  Main text |  Table 1 |  - |  FMT results - merge (donor, rec. pre- and post-FMT), odds ratio, p-value, cutoff as AMPs/MAG |  Done |  ftests.py |  
 |  Table |  Supp. |  Table S1 |  - |  Metagenome and Genome Samples in AMPSphere (sample, habitat, raw inserts, assembled bp, N50, ORFs+smORFs, smORFs, non-redundant AMPs) |  Done |  suptable_s1.py |  
 |  Table |  Supp. |  Table S2 |  - |  MannWhitney U test of normalized AMPs per assembled Mbp in different mammalian guts (Species 1, Species 2, U, P-value) |  Done |  plot-metadata.py |  
 |  Text |  Supp. |  Supp. Info S1 |  - |  Quality assessment of AMPSphere and its homologs |  Done |  qualtest.py |  
