{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8568898d",
   "metadata": {},
   "source": [
    "# AMPSphere v.2022-03\n",
    "\n",
    "This is a notebook meant to form the set of notebooks used to analyze the data in AMPSphere and write the manuscript:\n",
    "\n",
    "__AMPSphere: Global survey of prokaryotic antimicrobial peptides shaping microbiomes__\n",
    "\n",
    "### Check the c_AMP density of species through different environments controlling-quality of AMPs through the coordinates test\n",
    "\n",
    "We generate sets of c_AMPs by species and sample excluding those peptides that did not pass the coordinates test. Then group them by environment and select those species happening in at least 2 habitats with at least 10 samples in each. The number of non-redundant c_AMPs attributed to a given species per sample in each sample is compared using Mann-Whitney U test and later the p-values are adjusted using Holm-Sidak."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8fb71654",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from itertools import combinations, chain\n",
    "from scipy.stats import spearmanr, mannwhitneyu\n",
    "from statsmodels.stats.multitest import multipletests as mt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d5707b3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "data = pd.read_table('../data_folder/complete_amps_associated_taxonomy.tsv.gz')\n",
    "\n",
    "metadata = pd.read_table(\"../data_folder/metadata.tsv.xz\")\n",
    "metadata = metadata.rename({'sample_accession': 'sample'}, axis=1)\n",
    "\n",
    "quality = pd.read_table('../data_folder/quality_assessment.tsv.xz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ec494ebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# preparing data\n",
    "quality.set_index('AMP', inplace=True)\n",
    "quality = list(quality[quality.Coordinates == 'Passed'].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "03b47dbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter species-specific genes\n",
    "data = data[data.amp.isin(quality)]\n",
    "data = data.loc[data.level == 'species', ['sample', 'source']]\n",
    "data = data.groupby(['sample', 'source']).agg('size')\n",
    "data = data.reset_index()\n",
    "data = data.rename({'source': 'name',\n",
    "                    0: 'amp_genes'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "173962e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "genenvo = pd.read_table('../data_folder/general_envo_names.tsv.xz')\n",
    "\n",
    "metadata = metadata.merge(on=['microontology', 'host_scientific_name', 'host_tax_id'],\n",
    "                          right=genenvo,\n",
    "                          how='outer')\n",
    "\n",
    "metadata = metadata[['sample', 'general_envo_name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ee6d94fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate densities\n",
    "sps = pd.read_table('../data_folder/bps-per-sample-per-species.tsv.xz',\n",
    "                    sep='\\t',\n",
    "                    header='infer')\n",
    "\n",
    "data = data.merge(on=['sample', 'name'], right=sps)\n",
    "\n",
    "data['amp_density'] = data['amp_genes'] * 1e6 / data['nbps']\n",
    "\n",
    "# add metadata\n",
    "data = data.merge(on='sample', right=metadata)\n",
    "data.general_envo_name = data.general_envo_name.replace('human saliva', 'human mouth')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7e7b5a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tukeyfence(df):\n",
    "    itdf = pd.DataFrame()\n",
    "    for record in df.groupby(['name', 'general_envo_name']):\n",
    "        record = record[1]\n",
    "        if len(record) > 9:\n",
    "            q1, q3 = record['amp_density'].quantile([0.25,0.75])\n",
    "            iqr = q3-q1\n",
    "            ul, ll = q3+(1.5*iqr), q1-(1.5*iqr)\n",
    "            record = record[(record.amp_density <= ul) & (record.amp_density >= ll)]\n",
    "            itdf = pd.concat([itdf, record])\n",
    "    return itdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7663382d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>name</th>\n",
       "      <th>amp_genes</th>\n",
       "      <th>taxid</th>\n",
       "      <th>nbps</th>\n",
       "      <th>amp_density</th>\n",
       "      <th>general_envo_name</th>\n",
       "      <th>moe</th>\n",
       "      <th>moe_pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>773023</th>\n",
       "      <td>SAMN06264886</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>2</td>\n",
       "      <td>44747</td>\n",
       "      <td>1128853</td>\n",
       "      <td>1.771710</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>2.455462</td>\n",
       "      <td>138.592806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789120</th>\n",
       "      <td>SAMN06266445</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>1</td>\n",
       "      <td>44747</td>\n",
       "      <td>767610</td>\n",
       "      <td>1.302745</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>2.553378</td>\n",
       "      <td>195.999872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835246</th>\n",
       "      <td>SAMN06344006</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>1</td>\n",
       "      <td>44747</td>\n",
       "      <td>613652</td>\n",
       "      <td>1.629588</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>3.193990</td>\n",
       "      <td>195.999840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972177</th>\n",
       "      <td>SAMN07960327</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>2</td>\n",
       "      <td>44747</td>\n",
       "      <td>1089727</td>\n",
       "      <td>1.835322</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>2.543624</td>\n",
       "      <td>138.592802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972314</th>\n",
       "      <td>SAMN07960328</td>\n",
       "      <td>0-14-0-20-30-16 sp002779075</td>\n",
       "      <td>2</td>\n",
       "      <td>44747</td>\n",
       "      <td>1100407</td>\n",
       "      <td>1.817509</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>2.518937</td>\n",
       "      <td>138.592803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1373903</th>\n",
       "      <td>TARA_142_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>1</td>\n",
       "      <td>1762</td>\n",
       "      <td>4485880</td>\n",
       "      <td>0.222922</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.436926</td>\n",
       "      <td>195.999978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1374056</th>\n",
       "      <td>TARA_145_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>1</td>\n",
       "      <td>1762</td>\n",
       "      <td>4258030</td>\n",
       "      <td>0.234850</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.460307</td>\n",
       "      <td>195.999977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1374405</th>\n",
       "      <td>TARA_148b_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>3</td>\n",
       "      <td>1762</td>\n",
       "      <td>3802238</td>\n",
       "      <td>0.789009</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.892847</td>\n",
       "      <td>113.160608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1374468</th>\n",
       "      <td>TARA_149_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>1</td>\n",
       "      <td>1762</td>\n",
       "      <td>3801975</td>\n",
       "      <td>0.263021</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.515521</td>\n",
       "      <td>195.999974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1374967</th>\n",
       "      <td>TARA_152_MES_0.22-3_G</td>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>1</td>\n",
       "      <td>1762</td>\n",
       "      <td>3417823</td>\n",
       "      <td>0.292584</td>\n",
       "      <td>marine</td>\n",
       "      <td>0.573464</td>\n",
       "      <td>195.999971</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1133749 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sample                         name  amp_genes  \\\n",
       "773023             SAMN06264886  0-14-0-20-30-16 sp002779075          2   \n",
       "789120             SAMN06266445  0-14-0-20-30-16 sp002779075          1   \n",
       "835246             SAMN06344006  0-14-0-20-30-16 sp002779075          1   \n",
       "972177             SAMN07960327  0-14-0-20-30-16 sp002779075          2   \n",
       "972314             SAMN07960328  0-14-0-20-30-16 sp002779075          2   \n",
       "...                         ...                          ...        ...   \n",
       "1373903   TARA_142_MES_0.22-3_G        Zunongwangia profunda          1   \n",
       "1374056   TARA_145_MES_0.22-3_G        Zunongwangia profunda          1   \n",
       "1374405  TARA_148b_MES_0.22-3_G        Zunongwangia profunda          3   \n",
       "1374468   TARA_149_MES_0.22-3_G        Zunongwangia profunda          1   \n",
       "1374967   TARA_152_MES_0.22-3_G        Zunongwangia profunda          1   \n",
       "\n",
       "         taxid     nbps  amp_density general_envo_name       moe     moe_pct  \n",
       "773023   44747  1128853     1.771710       groundwater  2.455462  138.592806  \n",
       "789120   44747   767610     1.302745       groundwater  2.553378  195.999872  \n",
       "835246   44747   613652     1.629588       groundwater  3.193990  195.999840  \n",
       "972177   44747  1089727     1.835322       groundwater  2.543624  138.592802  \n",
       "972314   44747  1100407     1.817509       groundwater  2.518937  138.592803  \n",
       "...        ...      ...          ...               ...       ...         ...  \n",
       "1373903   1762  4485880     0.222922            marine  0.436926  195.999978  \n",
       "1374056   1762  4258030     0.234850            marine  0.460307  195.999977  \n",
       "1374405   1762  3802238     0.789009            marine  0.892847  113.160608  \n",
       "1374468   1762  3801975     0.263021            marine  0.515521  195.999974  \n",
       "1374967   1762  3417823     0.292584            marine  0.573464  195.999971  \n",
       "\n",
       "[1133749 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove outliers\n",
    "df = tukeyfence(data)\n",
    "df['p'] = df.amp_density / 1e6\n",
    "df['moe'] = np.sqrt(df.p*(1-df.p)/df.nbps) * 1.96 * 1e6\n",
    "df.drop('p', axis=1, inplace=True)\n",
    "df['moe_pct'] = df.moe * 100 / df.amp_density\n",
    "df.to_csv('species_amp_density_per_sample_hq.tsv.gz',\n",
    "          sep='\\t',\n",
    "          header=True,\n",
    "          index=None)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1653ed08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select species present in >=10 samples per habitats\n",
    "k = df.groupby(['name', 'general_envo_name']).agg('size')\n",
    "k = k[k >= 10]\n",
    "\n",
    "# select species present in more than 1 habitat\n",
    "k = k.reset_index().groupby('name').apply(lambda x: x.general_envo_name.tolist())\n",
    "k = k[k.apply(lambda x: len(x)) > 1]\n",
    "k = k.reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30ccefdf",
   "metadata": {},
   "source": [
    "### Densities per species and habitats controlling quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2f719233",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/celio/miniconda3/lib/python3.9/site-packages/statsmodels/stats/multitest.py:177: RuntimeWarning: divide by zero encountered in log1p\n",
      "  np.log1p(-pvals))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>env1</th>\n",
       "      <th>env1_samples</th>\n",
       "      <th>amp_density_avg_e1</th>\n",
       "      <th>amp_density_std_e1</th>\n",
       "      <th>env2</th>\n",
       "      <th>env2_samples</th>\n",
       "      <th>amp_density_avg_e2</th>\n",
       "      <th>amp_density_std_e2</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>25</td>\n",
       "      <td>0.997724</td>\n",
       "      <td>0.426723</td>\n",
       "      <td>marine</td>\n",
       "      <td>21</td>\n",
       "      <td>0.944880</td>\n",
       "      <td>0.611923</td>\n",
       "      <td>4.402062e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>25</td>\n",
       "      <td>0.997724</td>\n",
       "      <td>0.426723</td>\n",
       "      <td>soil</td>\n",
       "      <td>132</td>\n",
       "      <td>1.095878</td>\n",
       "      <td>0.785271</td>\n",
       "      <td>7.789809e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>groundwater</td>\n",
       "      <td>25</td>\n",
       "      <td>0.997724</td>\n",
       "      <td>0.426723</td>\n",
       "      <td>water associated</td>\n",
       "      <td>42</td>\n",
       "      <td>0.702492</td>\n",
       "      <td>0.341929</td>\n",
       "      <td>7.426659e-03</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>marine</td>\n",
       "      <td>21</td>\n",
       "      <td>0.944880</td>\n",
       "      <td>0.611923</td>\n",
       "      <td>soil</td>\n",
       "      <td>132</td>\n",
       "      <td>1.095878</td>\n",
       "      <td>0.785271</td>\n",
       "      <td>5.723084e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0-14-0-80-60-11 sp002779455</td>\n",
       "      <td>marine</td>\n",
       "      <td>21</td>\n",
       "      <td>0.944880</td>\n",
       "      <td>0.611923</td>\n",
       "      <td>water associated</td>\n",
       "      <td>42</td>\n",
       "      <td>0.702492</td>\n",
       "      <td>0.341929</td>\n",
       "      <td>1.682533e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7151</th>\n",
       "      <td>Zag111 sp004558955</td>\n",
       "      <td>human gut</td>\n",
       "      <td>53</td>\n",
       "      <td>163.041577</td>\n",
       "      <td>91.571776</td>\n",
       "      <td>pig gut</td>\n",
       "      <td>133</td>\n",
       "      <td>3.930803</td>\n",
       "      <td>3.142374</td>\n",
       "      <td>2.100328e-26</td>\n",
       "      <td>1.484512e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7152</th>\n",
       "      <td>Zag111 sp004558955</td>\n",
       "      <td>human gut</td>\n",
       "      <td>53</td>\n",
       "      <td>163.041577</td>\n",
       "      <td>91.571776</td>\n",
       "      <td>primate gut</td>\n",
       "      <td>15</td>\n",
       "      <td>21.175229</td>\n",
       "      <td>18.581715</td>\n",
       "      <td>9.540071e-09</td>\n",
       "      <td>5.839307e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7153</th>\n",
       "      <td>Zag111 sp004558955</td>\n",
       "      <td>pig gut</td>\n",
       "      <td>133</td>\n",
       "      <td>3.930803</td>\n",
       "      <td>3.142374</td>\n",
       "      <td>primate gut</td>\n",
       "      <td>15</td>\n",
       "      <td>21.175229</td>\n",
       "      <td>18.581715</td>\n",
       "      <td>8.391519e-05</td>\n",
       "      <td>3.071159e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7154</th>\n",
       "      <td>Zoogloea ramigera</td>\n",
       "      <td>activated sludge</td>\n",
       "      <td>61</td>\n",
       "      <td>0.660189</td>\n",
       "      <td>0.353442</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>116</td>\n",
       "      <td>0.685994</td>\n",
       "      <td>0.538758</td>\n",
       "      <td>2.765764e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7155</th>\n",
       "      <td>Zoogloea sp002028455</td>\n",
       "      <td>activated sludge</td>\n",
       "      <td>49</td>\n",
       "      <td>0.943846</td>\n",
       "      <td>0.436325</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>76</td>\n",
       "      <td>2.044889</td>\n",
       "      <td>2.242145</td>\n",
       "      <td>1.104769e-02</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7156 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          species              env1  env1_samples  \\\n",
       "0     0-14-0-80-60-11 sp002779455       groundwater            25   \n",
       "1     0-14-0-80-60-11 sp002779455       groundwater            25   \n",
       "2     0-14-0-80-60-11 sp002779455       groundwater            25   \n",
       "3     0-14-0-80-60-11 sp002779455            marine            21   \n",
       "4     0-14-0-80-60-11 sp002779455            marine            21   \n",
       "...                           ...               ...           ...   \n",
       "7151           Zag111 sp004558955         human gut            53   \n",
       "7152           Zag111 sp004558955         human gut            53   \n",
       "7153           Zag111 sp004558955           pig gut           133   \n",
       "7154            Zoogloea ramigera  activated sludge            61   \n",
       "7155         Zoogloea sp002028455  activated sludge            49   \n",
       "\n",
       "      amp_density_avg_e1  amp_density_std_e1              env2  env2_samples  \\\n",
       "0               0.997724            0.426723            marine            21   \n",
       "1               0.997724            0.426723              soil           132   \n",
       "2               0.997724            0.426723  water associated            42   \n",
       "3               0.944880            0.611923              soil           132   \n",
       "4               0.944880            0.611923  water associated            42   \n",
       "...                  ...                 ...               ...           ...   \n",
       "7151          163.041577           91.571776           pig gut           133   \n",
       "7152          163.041577           91.571776       primate gut            15   \n",
       "7153            3.930803            3.142374       primate gut            15   \n",
       "7154            0.660189            0.353442        wastewater           116   \n",
       "7155            0.943846            0.436325        wastewater            76   \n",
       "\n",
       "      amp_density_avg_e2  amp_density_std_e2       p_value         p_adj  \n",
       "0               0.944880            0.611923  4.402062e-01  1.000000e+00  \n",
       "1               1.095878            0.785271  7.789809e-01  1.000000e+00  \n",
       "2               0.702492            0.341929  7.426659e-03  1.000000e+00  \n",
       "3               1.095878            0.785271  5.723084e-01  1.000000e+00  \n",
       "4               0.702492            0.341929  1.682533e-01  1.000000e+00  \n",
       "...                  ...                 ...           ...           ...  \n",
       "7151            3.930803            3.142374  2.100328e-26  1.484512e-22  \n",
       "7152           21.175229           18.581715  9.540071e-09  5.839307e-05  \n",
       "7153           21.175229           18.581715  8.391519e-05  3.071159e-01  \n",
       "7154            0.685994            0.538758  2.765764e-01  1.000000e+00  \n",
       "7155            2.044889            2.242145  1.104769e-02  1.000000e+00  \n",
       "\n",
       "[7156 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test difference in the amp_densitys among samples of\n",
    "# the same species through different environments\n",
    "test = []\n",
    "for _, s, env in k.itertuples():\n",
    "    combs = combinations(env, 2)\n",
    "    for i, j in combs:\n",
    "        n1 = df[(df.name == s) & (df.general_envo_name == i)]\n",
    "        n2 = df[(df.name == s) & (df.general_envo_name == j)]\n",
    "        u, p = mannwhitneyu(n1['amp_density'], n2['amp_density'])\n",
    "        test.append((s, i, len(n1),\n",
    "                     n1.amp_density.mean(), n1.amp_density.std(),\n",
    "                     j, len(n2), n2.amp_density.mean(),\n",
    "                     n2.amp_density.std(), p))\n",
    "\n",
    "# format results\n",
    "test = pd.DataFrame(test, columns=['species',\n",
    "                                   'env1',\n",
    "                                   'env1_samples',\n",
    "                                   'amp_density_avg_e1',\n",
    "                                   'amp_density_std_e1',\n",
    "                                   'env2',\n",
    "                                   'env2_samples',\n",
    "                                   'amp_density_avg_e2',\n",
    "                                   'amp_density_std_e2',\n",
    "                                   'p_value'])\n",
    "\n",
    "test\n",
    "\n",
    "# adjust p-values\n",
    "_, test['p_adj'], _, _ = mt(test['p_value'], method='hs')\n",
    "\n",
    "# export\n",
    "test.to_csv('species_amp_density_crossenvironment_hq.tsv.gz', sep='\\t', header=True, index=None)\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7e5281a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "It was tested 2995 species\n",
      "Only 39.93% (1196) of them\n",
      "presented significant differences in their AMP density\n",
      "across different environments\n",
      "In average, it was observed 2.4e+01 c_AMP genes per assembled Mbp for each species\n",
      "per sample, ranging between 2.329e+03-0.12.\n"
     ]
    }
   ],
   "source": [
    "### Generating statistics\n",
    "\n",
    "# select environments with at least 100 samples containing AMPs\n",
    "k = data[['sample', 'general_envo_name']].drop_duplicates()\n",
    "k = k.general_envo_name.value_counts()\n",
    "k = k[k>=100].index\n",
    "\n",
    "# select pairs containing both environments in the list\n",
    "x = test[(test.env1.isin(k)) & (test.env2.isin(k))]\n",
    "\n",
    "n1 = len(set(x.species))\n",
    "print(f'It was tested {n1} species')\n",
    "\n",
    "n2 = len(set(x.loc[test.p_adj < 0.05, 'species']))\n",
    "print(f'''Only {round((n2*100/n1), 2)}% ({n2}) of them\n",
    "presented significant differences in their AMP density\n",
    "across different environments''')\n",
    "\n",
    "savg = (x.amp_density_avg_e1.sum() + x.amp_density_avg_e2.sum()) / (2*len(x))\n",
    "\n",
    "mmax, mmin = [max(x.amp_density_avg_e1.max(), x.amp_density_avg_e2.max()),\n",
    "              min(x.amp_density_avg_e1.min(), x.amp_density_avg_e2.min())]\n",
    "\n",
    "print(f'''In average, it was observed {savg:.2} c_AMP genes per assembled Mbp for each species\n",
    "per sample, ranging between {mmax:.4}-{mmin:.2}.''')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
